src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx

```diff
@@ -0,0 +1,368 @@
+import {observer} from "mobx-react";
+import * as React from "react";
+import {VictoryChart, VictorySelectionContainer, VictoryAxis, VictoryLabel, VictoryScatter} from "victory";
+import CBIOPORTAL_VICTORY_THEME from "../../../../shared/theme/cBioPoralTheme";
+import {computed, observable} from "mobx";
+import autobind from "autobind-decorator";
+import {tickFormatNumeral} from "../../../../shared/components/plots/TickUtils";
+import {makeMouseEvents} from "../../../../shared/components/plots/PlotUtils";
+import _ from "lodash";
+import {downsampleByGrouping, DSData} from "../../../../shared/components/plots/downsampleByGrouping";
+import ScatterPlotTooltip from "../../../../shared/components/plots/ScatterPlotTooltip";
+import {DOWNSAMPLE_PIXEL_DISTANCE_THRESHOLD, getDownsampledData, MAX_DOT_SIZE} from "./StudyViewScatterPlotUtils";
+import {ClinicalAttribute, SampleIdentifier} from "../../../../shared/api/generated/CBioPortalAPI";
+import LoadingIndicator from "shared/components/loadingIndicator/LoadingIndicator"
+import $ from "jquery";
+import {AnalysisGroup} from "../../StudyViewPageStore";
+import {AbstractChart} from "../ChartContainer";
+import {interpolatePlasma} from "d3-scale-chromatic";
+
+export interface IStudyViewDensityScatterPlotData {
+    x:number;
+    y:number;
+    uniqueSampleKey:string;
+    studyId:string;
+    sampleId:string;
+    patientId:string;
+}
+
+export interface IStudyViewDensityScatterPlotProps {
+    width:number;
+    height:number;
+    data:IStudyViewDensityScatterPlotData[]
+    onSelection:(sampleIdentifiers:SampleIdentifier[], keepCurrent:boolean)=>void;
+
+    isLoading?:boolean;
+    svgRef?:(svg:SVGElement|null)=>void;
+    tooltip?:(d:DSData<IStudyViewDensityScatterPlotData>)=>JSX.Element;
+    axisLabelX?: string;
+    axisLabelY?: string;
+    title?:string;
+}
+
+const NUM_AXIS_TICKS = 8;
+const DOMAIN_PADDING = 15;
+const BIN_THRESHOLD = 0;
+
+@observer
+export default class StudyViewDensityScatterPlot extends React.Component<IStudyViewDensityScatterPlotProps, {}> implements AbstractChart {
+    @observable tooltipModel:any|null = null;
+    @observable pointHovered:boolean = false;
+    @observable mouseIsDown:boolean = false;
+    @observable shiftPressed:boolean = false;
```

Comment:
When keepCurrent is set to true, the shiftPressed event should not be used

Reply:
shiftPressed determines whether keepCurrent is used. I'm not sure what you're saying

Reply:
NM. I misunderstood the meaning of the shiftclick.